primo/utils/clustering_utils.py

```diff
@@ -145,3 +177,106 @@ def perform_clustering(wd: WellData, distance_threshold: float = 10.0):
     wd.col_names.register_new_columns({"cluster": "Clusters"})
 
     return clustered_data.n_clusters_
+
+
+def perform_louvain_clustering(
+    wd: WellData, distance_threshold: float = 10.0, cluster_threshold: int = 300
+) -> dict:
+    """
+    Partitions the data into smaller clusters using the Louvain community detection method,
+    limiting each well to connect only with its 10 closest neighbors within a threshold distance.
+    Dynamically adjusts the resolution parameter to ensure no cluster exceeds the size threshold.
+
+    Parameters
+    ----------
+    wd : WellData
+        Object containing well data
+
+    distance_threshold : float
+        Threshold distance (in miles) for breaking clusters
+
+    cluster_threshold : int
+        cluster size of the largest cluster
+
+    Returns
+    -------
+    dict
+        Dictionary of lists of wells contained in each cluster
+    """
+    # pylint: disable=too-many-locals
+    existing_clusters = _check_existing_cluster(wd)
+    if existing_clusters is not None:
+        return existing_clusters
+
+    coordinates = list(
+        zip(wd.data[wd.col_names.latitude], wd.data[wd.col_names.longitude])
+    )
+    ball_tree = BallTree(np.radians(coordinates), metric="haversine")
+
+    distances, indices = ball_tree.query(np.radians(coordinates), k=11)
```

Comment:
Can we not account for 'age' and 'depth' in our distance calculations? That is a must in my opinion

Reply:
Definitely can, although some more work will need to go into balancing out the different metrics. This also might give overlapping clusters, which I think we decided is ok (although if two overlapping clusters are both selected as projects, the results might not look "right"?). It looks like it mirrors the current agglomerative clustering method (line 164) weights.

Reply:
Note my comment is referring to Louvain itself. In actually constructing the graph, I am not sure.

Reply:
The question here is: do we construct the graph by distance and these other metrics, or do we partition the graph based on distance and these other metrics, or both?